db/db_impl/db_impl_write.cc

```diff
@@ -1737,6 +1737,178 @@ void DBImpl::NotifyOnMemTableSealed(ColumnFamilyData* /*cfd*/,
+      new_earliest_seqno = ikey.sequence < new_earliest_seqno
+                               ? ikey.sequence
+                               : new_earliest_seqno;
+      new_first_seqno =
+          ikey.sequence < new_first_seqno ? ikey.sequence : new_first_seqno;
+      new_mem->SetFirstSequenceNumber(new_first_seqno);
+
+      // Should we update "OldestKeyTime" ????
+      s = new_mem->Add(
+          ikey.sequence, ikey.type, ikey.user_key, value,
+          nullptr,   // KV protection info set as nullptr since it
+                     // should only be useful for the first add to
+                     // the original memtable.
+          false,     // : allow concurrent_memtable_writes_
+                     // Not seen as necessary for now.
+          nullptr,   // get_post_process_info(m) must be nullptr
+                     // when concurrent_memtable_writes is switched off.
+          nullptr);  // hint, only used when concurrent_memtable_writes_
+                     // is switched on.
+      if (!s.ok()) {
+        break;
+      }
+    }
+
+    // Check status and propagate
+    // potential error status from c_iter
+    if (!s.ok()) {
+      c_iter.status().PermitUncheckedError();
+    } else if (!c_iter.status().ok()) {
+      s = c_iter.status();
+    }
+
+    // Range tombstone transfer.
+    if (s.ok()) {
+      auto range_del_it = range_del_agg->NewIterator();
+      for (range_del_it->SeekToFirst(); range_del_it->Valid();
+           range_del_it->Next()) {
+        auto tombstone = range_del_it->Tombstone();
+        new_first_seqno =
+            tombstone.seq_ < new_first_seqno ? tombstone.seq_ : new_first_seqno;
+        new_mem->SetFirstSequenceNumber(new_first_seqno);
+        s = new_mem->Add(
+            tombstone.seq_,        // Sequence number
+            kTypeRangeDeletion,    // KV type
+            tombstone.start_key_,  // Key is start key.
+            tombstone.end_key_,    // Value is end key.
+            nullptr,               // KV protection info set as nullptr since it
+                                   // should only be useful for the first add to
+                                   // the original memtable.
+            false,                 // : allow concurrent_memtable_writes_
+                                   // Not seen as necessary for now.
+            nullptr,               // get_post_process_info(m) must be nullptr
+                      // when concurrent_memtable_writes is switched off.
+            nullptr);  // hint, only used when concurrent_memtable_writes_
+                       // is switched on.
+
+        if (!s.ok()) {
+          break;
+        }
+      }
+    }
+  }
+  // Note: if the mempurge was ineffective, meaning that there was no
+  // garbage to remove, and this new_mem needs to be flushed again,
+  // the new_mem->Add would have updated the flush status when it
+  // called "UpdateFlushState()" internally at the last Add() call.
```

Comment:
I'm not clear on the flow if the mempurge was ineffective. How do we prevent an endless loop of purge attempts of the same memtable data?

Reply:
You are right, I forgot to add an error status. 
What is happening now is at the end of the MemPurge function I check if the new memtable needs to be flushed again. If this is the case, I change the return status such that when the MemPurge function returns, the error status is noticed, the mempurge is aborted, and the switchMemtable process is rerouted to a regular flush operation (see db/db_impl/db_impl_write.cc:2119).